src/rgw/rgw_lua.cc

```diff
@@ -135,5 +139,118 @@ int delete_script(rgw::sal::RGWRadosStore* store, const std::string& tenant, opt
 
   return 0;
 }
+
+#ifdef WITH_RADOSGW_LUA_MODULES
+
+const std::string MODULE_LIST_OBJECT_NAME = "lua_modules_allowlist";
+
+namespace bp = boost::process;
+
+int add_module(rgw::sal::RGWRadosStore* store, optional_yield y, const std::string& module_name, bool allow_compilation) {
+  // verify that luarocks can load this module
+  const auto p = bp::search_path("luarocks");
+  if (p.empty()) {
+    return -ECHILD;
+  }
+  bp::ipstream is;
+  const auto cmd = p.string() + " search --porcelain" + (allow_compilation ? " " : " --binary ") + module_name;
+  bp::child c(cmd,
+      bp::std_in.close(),
+      bp::std_err > bp::null,
+      bp::std_out > is);
+
+  std::string line;
+  bool module_found = false;
+  // TODO: yield on reading the output
+  while (c.running() && std::getline(is, line) && !line.empty()) {
+    module_found = true;
+  }
+  c.wait();
+  auto ret = c.exit_code();
+  if (ret) {
+    return -ret;
+  }
+
+  if (!module_found) {
+    return -EINVAL;
+  }
+  
+  // add module to list
+  const bufferlist empty_bl;
+  std::map<std::string, bufferlist> new_module{{module_name, empty_bl}};
+  librados::ObjectWriteOperation op;
+  op.omap_set(new_module);
+  ret = rgw_rados_operate(*(store->getRados()->get_lc_pool_ctx()), 
+      MODULE_LIST_OBJECT_NAME, &op, y);
+
+  if (ret < 0) {
+    return ret;
+  } 
+  return 0;
+}
+
+int list_modules(rgw::sal::RGWRadosStore* store, optional_yield y, modules_t& modules) {
+  constexpr auto max_chunk = 1024U;
+  std::string start_after;
+  bool more = true;
+  int rval;
+  while (more) {
+    librados::ObjectReadOperation op;
+    modules_t modules_chunk;
+    op.omap_get_keys2(start_after, max_chunk, &modules_chunk, &more, &rval);
+    const auto ret = rgw_rados_operate(*(store->getRados()->get_lc_pool_ctx()),
+      MODULE_LIST_OBJECT_NAME, &op, nullptr, y);
+  
+    if (ret < 0) {
+      return ret;
+    }
+
+    modules.merge(modules_chunk);
+  }
+ 
+  return 0;
+}
+
+int install_modules(rgw::sal::RGWRadosStore* store, optional_yield y, modules_t& failed_modules, std::string& output) {
+  modules_t modules;
+  auto ret = list_modules(store, y, modules);
+  if (ret == -ENOENT) {
+    // allowlist is empty 
+    return 0;
+  }
+  if (ret < 0) {
+    return ret;
+  }
+  // verify that luarocks can load this module
+  const auto p = bp::search_path("luarocks");
+  if (p.empty()) {
+    return -ECHILD;
+  }
+  // LUA_MODULES_INSTALL_DIR will be created by luarocks the first time it is called
+  std::for_each(modules.begin(), modules.end(), [&p, &failed_modules, &output](const std::string& module_name) {
+    bp::ipstream is;
+    bp::child c(p, "install", "--lua-version", "5.3", "--tree", LUA_MODULES_INSTALL_DIR, "--deps-mode", "one", module_name, 
```

Comment:
might want to use `find_package(Lua)` for its version number.

Reply:
removing the version. since now `FindLua()` may get a version newer than 5.3